### PR TITLE
Option to Loss Exp When Die

### PR DESCRIPTION
**Pull request by dhmhc**
Resolves: #722

### CODE REVIEW

File: Intersect (Core)/Config/PlayerOptions.cs [hunk 1]
```diff
@@ -59,6 +59,12 @@ public class PlayerOptions
         /// </summary>
         public bool ShowLevelByName = false;
 
+        /// <summary>
+        /// First thanks for letting me share an idea, of losing XP when dying <3
+        /// </summary>
+        public bool XPLossOnDeath = false;
```
Comment (jcsnider) [inline Intersect (Core)/Config/PlayerOptions.cs:65]: We don't need a boolean here.

File: Intersect (Core)/Config/PlayerOptions.cs [hunk 2]
```diff
@@ -59,6 +59,12 @@ public class PlayerOptions
...
+        public int XPLossPercent = 0;
```
Comment (jcsnider) [inline Intersect (Core)/Config/PlayerOptions.cs:66]: Rename this to ExpLossOnDeathPercent

File: Intersect.Server/Entities/Player.cs [hunk 3]
```diff
@@ -812,7 +812,13 @@ public override void Die(bool dropItems = true, Entity killer = null)
             {
                 base.Die(dropItems, killer);
             }
-            
+
+            if (Options.Player.XPLossOnDeath.Equals(true))
```
Comment (jcsnider) [inline Intersect.Server/Entities/Player.cs:816]: Then do

if (Options.Instance.Player.ExpLossOnDeathPercent > 0)

File: Intersect.Server/Entities/Player.cs [hunk 4]
```diff
@@ -812,7 +812,13 @@ public override void Die(bool dropItems = true, Entity killer = null)
...
+            {
+                var ExpLoss = (GetExperienceToNextLevel(this.Level) * (Options.Player.XPLossPercent / 100.0));
```
Comment (jcsnider) [inline Intersect.Server/Entities/Player.cs:818]: This isn't calculating based on the current xp...

instead you want:
var ExpLoss = (this.Exp * (Options.Player.XPLossPercent / 100.0));

**Review by jcsnider — CHANGES_REQUESTED** [review]

File: Intersect.Server/Entities/Player.cs (hunk 4 above)
Comment (dhmhc) [reply Intersect.Server/Entities/Player.cs:818]: But the XP loss is calculated based on the total required to level and not the xp that the player currently has, that was purposeful, I believe that in most games the XP loss calculation is the percentage over what he needs to upar and not what he already has, no?

Comment (jcsnider) [reply Intersect.Server/Entities/Player.cs]: I might be wrong I but I don't think that's how everyone will want it.... how about one more boolean in the options?

public bool ExpLossFromCurrentExp {get; set;} = true

Then you can do an if/then to calculate exp loss based on both?

Comment (dhmhc) [reply Intersect.Server/Entities/Player.cs]: Great idea I will change the files here, and let the staff choose the mode

Comment (dhmhc) [reply Intersect.Server/Entities/Player.cs]: I think its all ok now 👯

**Comment by dhmhc** [comment]
Any further changes needed? 🍰

File: Intersect (Core)/Config/PlayerOptions.cs [hunk 5]
```diff
@@ -59,6 +59,15 @@ public class PlayerOptions
         /// </summary>
         public bool ShowLevelByName = false;
 
+        /// <summary>
+        /// First thanks for letting me share an idea, of losing XP when dying <3
+        /// </summary>
+        public int ExpLossOnDeathPercent = 0;
+        /// <summary>
```
Comment (jcsnider) [inline Intersect (Core)/Config/PlayerOptions.cs:66]: Please add a a line break before this summary so it matches the rest of the file aesthetically.

File: Intersect.Server/Entities/Player.cs [hunk 6]
```diff
@@ -812,7 +812,20 @@ public override void Die(bool dropItems = true, Entity killer = null)
             {
                 base.Die(dropItems, killer);
             }
-            
+
+            if (Options.Instance.Player.ExpLossOnDeathPercent > 0)
+            {
+                if (Options.Player.ExpLossFromCurrentExp.Equals(true))
+                {
+                var ExpLoss = (this.Exp * (Options.Player.ExpLossOnDeathPercent / 100.0));
```
Comment (jcsnider) [inline Intersect.Server/Entities/Player.cs:820]: Please indent (add a tab) on these 2 lines.

File: Intersect.Server/Entities/Player.cs [hunk 7]
```diff
@@ -812,7 +812,20 @@ public override void Die(bool dropItems = true, Entity killer = null)
...
+                TakeExperience((long)ExpLoss);
+                }
+                else
+                {
+                var ExpLoss = (GetExperienceToNextLevel(this.Level) * (Options.Player.ExpLossOnDeathPercent / 100.0));
```
Comment (jcsnider) [inline Intersect.Server/Entities/Player.cs:825]: Please indent (add a tab) on these 2 lines.

File: Intersect.Server/Entities/Player.cs [hunk 8]
```diff
@@ -1010,6 +1023,16 @@ public void GiveExperience(long amount)
...
+        public void TakeExperience(long amount)
+        {
+            Exp -= amount;
+            if (Exp < 0)
+            {
+                Exp = 0;
+            }
+
+            PacketSender.SendExperience(this);
+        }
```
Comment (jcsnider) [inline Intersect.Server/Entities/Player.cs:1035]: Please add a line break after this function so there is space before the next.

**Review by jcsnider — CHANGES_REQUESTED** [review]
Added a few more comments to fix up some spacing/aesthetic issues. Once fixed I think this will be good to merge.

File: Intersect.Server/Entities/Player.cs (hunk 6 above)
Comment (dhmhc) [reply Intersect.Server/Entities/Player.cs]: Done

File: Intersect.Server/Entities/Player.cs (hunk 7 above)
Comment (dhmhc) [reply Intersect.Server/Entities/Player.cs]: Done

File: Intersect (Core)/Config/PlayerOptions.cs (hunk 5 above)
Comment (dhmhc) [reply Intersect (Core)/Config/PlayerOptions.cs:67]: Done

**Review by jcsnider — APPROVED** [review]

**Comment by Arufonsu** [comment]
@jcsnider i think that this should be moved to **Done**

**Pull request merged**